Support surface-based VecGeom 2.x navigator

mrguilima opened the pull request:
Adapting Celeritas to use the latest surface-based model from VecGeom 2.x  (see #1412).

The main change is to use VecGeom's `vgbrep::protonav::BVHSurfNavigator`. Since there are interface differences between this one and the volume-based `BVHNavigator`, the code-guards `VECGEOM_USE_SURF`, in analogy to AdePT.
Alternatively, `VECGEOM_VERSION >= 0x020000` could be used instead, as the VecGeom 1.2.x series should always be used for volume-based navigation anyway.

There are also distinct interfaces for ABBoxManager (template vs. non-template).

Note: in draft mode, as some tests currently fail.  Under investigation.

mrguilima committed:
Adapt to VG 2.x templated ABBoxManager

mrguilima committed:
Temporary fix to four-levels.gdml: replace unsupported orb with supported sphere

mrguilima committed:
Fixes to Vecgeom:FourLevelsTest*

mrguilima committed:
Updates to use the latest BVHSurfNavigator from VecGeom 2.x

Several tests hang, to be investigated

sethrj commented on src/geocel/vg/VecgeomParams.cc:
```diff
@@ -11,6 +11,7 @@
 #include <vector>
 #include <VecGeom/base/Config.h>
 #include <VecGeom/base/Cuda.h>
+#include <VecGeom/base/Version.h>
```
I've verified that this header (and the `VECGEOM_VERSION` macro therein) are present in v1.2.4, which is the minimum vecgeom required by Celeritas.

sethrj committed:
Revert "Fixes to Vecgeom:FourLevelsTest*"

This reverts commit 228ed5842dbe1fbe889f89d74918c7d0d70de3ef.

sethrj committed:
Add orb->sphere translation and disable unimplemented surface solids

sethrj commented on src/geocel/vg/detail/SurfNavigator.hh:
```diff
@@ -79,9 +88,9 @@ class SurfNavigator
-            globalpoint, globaldir, in_state, out_state, exit_surf, step_limit);
+	long hitsurf_id = 0;
+        auto step = vgbrep::protonav::BVHSurfNavigator<Precision>::ComputeStepAndNextSurface(
+            globalpoint, globaldir, in_state, out_state, hitsurf_id, step_limit);
```
Something here is misbehaving compared to the volume version:
```
54: /Users/seth/Code/celeritas-temp/test/geocel/vg/Vecgeom.test.cc:164: Failure
54: Value of: next.boundary
54:   Actual: true
54: Expected: false
```
for the calls
```
    auto next = geo.find_next_step(from_cm(100));
    EXPECT_SOFT_EQ(100, to_cm(next.distance));
    EXPECT_FALSE(next.boundary);
```

So it's claiming to have a boundary for the next surface, even though it doesn't hit it within the step limit.

sethrj committed:
Add surface geometry to CI

sethrj committed:
REVERTME: delete other CI builds

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vg-2.x

sethrj commented:
Without GPU on my laptop (clang 15 debug), on VecGeom ca34e35ea5fc822fcffd3c7c5d117e6bc671ce32 (reldeb), the following tests fail:
```
	 54 - geocel/vg/Vecgeom:SimpleCmsTest.* (Failed)
	 55 - geocel/vg/Vecgeom:FourLevelsGeantTest.* (Failed)
	 56 - geocel/vg/Vecgeom:SolidsGeantTest.* (SEGFAULT)
	 57 - geocel/vg/Vecgeom:ZnenvGeantTest.* (Failed)
	166 - celeritas/field/FieldPropagator (SEGFAULT)
	167 - celeritas/field/LinearPropagator (Failed)
	169 - celeritas/geo/Geometry (SEGFAULT)
	170 - celeritas/geo/GeoMaterial (Failed)
	171 - celeritas/global/AlongStep:-Em3*:SimpleCms*:LeadBox* (SEGFAULT)
	172 - celeritas/global/AlongStep:Em3AlongStepTest.nofluct_nomsc (Failed)
	173 - celeritas/global/AlongStep:Em3AlongStepTest.msc_nofluct (Failed)
	175 - celeritas/global/AlongStep:Em3AlongStepTest.fluct_nomsc (Failed)
	176 - celeritas/global/AlongStep:SimpleCmsAlongStepTest.msc_field (Subprocess killed)
	178 - celeritas/global/AlongStep:SimpleCmsRZFieldAlongStepTest.msc_rzfield (Failed)
	180 - celeritas/global/KernelContextException (Failed)
	181 - celeritas/global/Stepper (Failed)
	182 - celeritas/global/StepperGeant:-TestEm*:LeadBox* (Failed)
	183 - celeritas/global/StepperGeant:TestEm3Compton.* (Failed)
	184 - celeritas/global/StepperGeant:TestEm3NoMsc.* (Failed)
	185 - celeritas/global/StepperGeant:TestEm3Msc.* (Failed)
	186 - celeritas/global/StepperGeant:TestEm3MscNofluct.* (Failed)
	187 - celeritas/global/StepperGeant:TestEm15FieldMsc.* (Failed)
	203 - celeritas/optical/OpticalCollector (Failed)
	238 - celeritas/track/TrackSort (Failed)
	241 - celeritas/user/Diagnostic:-TestEm3* (Failed)
	242 - celeritas/user/Diagnostic:TestEm3* (Failed)
	243 - celeritas/user/StepCollector (SEGFAULT)
```

The simplest failure is just that it seems to report aa boundary when not actually hitting a boundary:
```
[ RUN      ] SimpleCmsTest.track
/Users/seth/Code/celeritas-temp/test/geocel/vg/Vecgeom.test.cc:164: Failure
Value of: next.boundary
  Actual: true
Expected: false

/Users/seth/Code/celeritas-temp/test/geocel/vg/Vecgeom.test.cc:176: Failure
Expected equality of these values:
  "si_tracker"
  this->volume_name(geo)
    Which is: "world"

[  FAILED  ] SimpleCmsTest.track (0 ms)
```

Full failure log:
[LastTest.log](https://github.com/user-attachments/files/17130055/LastTest.log)

sethrj committed:
Fix infinite loop in field propagator due to misbehaving boundary

sethrj committed:
Simplify initializer

sethrj committed:
Add two-boxes test and re-enable other vecgeom tests

sethrj committed:
Fix distance-to-boundary by initializing hit surface to -1

sethrj committed:
Revert orb-to-sphere conversion after vecgeom#1231

mrguilima committed:
Fix SurfNavigator for TwoBoxes*: new field, VecgeomTrackView::hit_surf_

mrguilima committed:
New reference values for CmseTest.trace

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vg-2.x

sethrj committed:
Always save cache if build succeeds

sethrj committed:
fixup! Always save cache if build succeeds

mrguilima committed:
Always reset hit_surf before calling finding next step

mrguilima committed:
No more warnings during loading of four-levels geometry

mrguilima committed:
No more warnings during loading of Cmse.gdml

sethrj commented on test/geocel/vg/Vecgeom.test.cc:
```diff
@@ -282,18 +365,19 @@ class FourLevelsTest : public VecgeomVgdmlTestBase
-    SpanStringView expected_log_levels() const final
-    {
-        if (vecgeom_version >= Version{2})
+    /*  SpanStringView expected_log_levels() const final
```
I suspect you mean this to be temporary, but it's a good habit to start using `#if 0` or (even better since you mean it *just* for the surface case)  use a conditional based on the surface settings macro.

sethrj commented on test/geocel/vg/Vecgeom.test.cc:
```diff
@@ -984,8 +1070,8 @@ TEST_F(CmseTest, trace)
-            0.52499999999986, 13.023518051922, 6.95, 6.95, 13.023518051922,
-            0.52499999999986, 2.15, 100, 5, 8, 100, 100, 100};
+            10.3027302206745, 13.023518051922, 6.95, 6.95, 13.023518051922,
+            10.3027302206745, 2.15, 100, 5, 8, 100, 100, 100};
```
It's great that you've recorded the updated safety values: these are now consistent with the ones returned by Geant4. I suppose we'll have to "edit" these when using the volume-based vecgeom.

sethrj commented on test/geocel/vg/Vecgeom.test.cc:
```diff
@@ -1034,7 +1120,8 @@ TEST_F(CmseTest, trace)
         EXPECT_VEC_SOFT_EQ(expected_distances, result.distances);
-        static real_type const expected_hw_safety[] = {1, 1, 242, 460};
+        static real_type const expected_hw_safety[] = {6.2475, 47.9499999999998,
+            242, 460};
```
Guess we'll have to change the label on this one 😄 since it now agrees between the two

mrguilima committed:
Disable tests related to navigation from the outside

sethrj committed:
Add CELERITAS_VECGEOM_SURFACE macro

sethrj committed:
Make next surface persistent

sethrj committed:
Loosen move-internal requirement for VecGeom

sethrj committed:
Remove warning expectation from vecgeom 2

sethrj committed:
Fix build

mrguilima committed:
Disabled two tests doing tracking outside of world

mrguilima committed:
GL temp: missing code

mrguilima committed:
Merge branch 'develop' into vg-2.x

mrguilima committed:
Fixes for VecGeom v2.x

mrguilima committed:
Switch test expected value depending on VecGeom version

mrguilima committed:
More temporary fixes to Vecgeom tests

mrguilima committed:
Merge branch 'develop' into vg-2.x

mrguilima committed:
Merge branch 'develop' into vg-2.x

mrguilima committed:
Extend CMake version to 3.31 and delay action on new CMP0175

mrguilima committed:
Extend CMake version to 3.31 and delay action on new CMP0175

mrguilima committed:
Fix Geometry test

mrguilima committed:
Fix expected volumes and safety values in Vecgeom.test.cc

mrguilima committed:
Fix expected volumes and safety values in Vecgeom.test.cc

mrguilima committed:
Merge branch 'develop' into vg-2.x

mrguilima committed:
Fix test Geometry:CmseTest*

mrguilima committed:
Add Solids tests based on solids model rather than VecGeom version

mrguilima committed:
Use BVHNavigator from VecGeom since VecGeom 2.0

mrguilima committed:
Adjust Vecgeom.test.cc expectations based on VecGeom version

mrguilima committed:
Merge branch 'temp3' into vg-2.x

mrguilima committed:
Fix SolidsGeantTest*

mrguilima committed:
Use VecGeom ABBoxManager alias when available

mrguilima committed:
Fix leftover from a merge conflict

mrguilima committed:
Avoid lots of tests in VecGeom surface model, until they get fixed

mrguilima committed:
Revert "Fix infinite loop in field propagator due to misbehaving boundary"

This reverts commit 97be3f7d2272c49520f211153bffa352cb17ad91.

mrguilima committed:
Merge branch 'develop' into vg-2.x

mrguilima committed:
Remove duplicate line

mrguilima committed:
Try to improve handling of relocation to [OUTSIDE]

mrguilima committed:
Putting back some geometry tests related to Orb

mrguilima committed:
Re-enable most tests to surface model after Orb fix in VecGeom

mrguilima committed:
Remove several test differences of surface model vs. solids model

mrguilima committed:
No need to call SurfNavigator::Relocate from [OUTSIDE]

VecgeomTrackView::find_next_step() properly sets the next state,
so cross_boundary() just needs to set vgstate_ = vgnext_

mrguilima committed:
Fix minor typo

mrguilima committed:
Split geo_Geometry test into three geometries, to avoid geometry reloading

mrguilima committed:
Split FieldPropagator tests to avoid loading multiple geometries per job

mrguilima committed:
Small corrections to documentation

mrguilima committed:
Fix typos in docs

mrguilima committed:
Fix to small difference between solid- and surface-model

mrguilima committed:
Add soft_zero() tolerance into linear_step to chord.length comparison

mrguilima committed:
Fix documentation

mrguilima committed:
Add extra step to ensure a test gamma reaches a boundary

mrguilima committed:
Account for Surface- vs. Solid-model difference

mrguilima committed:
Remove soft_zero(linear_step, chord.length) for good

mrguilima committed:
Merge branch 'develop' into vg-2.x

mrguilima committed:
Filter GeoMaterial tests by geometry

mrguilima committed:
Comment out some tests failing on surface model

mrguilima committed:
Relax tolerance for test failing on surface model

mrguilima committed:
Increase tolerances in Geometry test due to surface model issues

mrguilima committed:
Split Mock* test out of combined AlongStep testing

mrguilima committed:
Merge branch 'develop' into vg-2.x

sethrj committed:
Add runge-kutta citation

sethrj committed:
Fix documentation

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vg-2.x

sethrj committed:
Try fixing g4vg+surface on CI

sethrj committed:
Clean test

mrguilima committed:
Relax tolerances for the FourLevels tests

mrguilima committed:
Disable MultiLevelVgdml test due to unsupported reflection

mrguilima committed:
Relax safety tolerances for MultiLevel test

sethrj committed:
Unify safety tolerance

sethrj committed:
Fix doc

sethrj committed:
Fix cmse test on non-vecgeom geo

sethrj committed:
REVERTME: delete other CI builds

sethrj committed:
Use expressions instead of shell

sethrj committed:
Fix vgsurf configuration

sethrj committed:
Acutally use vgsurf

sethrj committed:
Fix preset

sethrj committed:
Update cmse tests

sethrj committed:
Attempt to fix configure

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vg-2.x

sethrj committed:
Make transitive geometry dependencies public

mrguilima committed:
Avoid #include for CUDA compilation

sethrj committed:
Make CELERITAS_VECGEOM_SURFACE always on

sethrj committed:
Update field propagator test

sethrj committed:
Loosen hepmc3 tolerance

sethrj committed:
Finish field propagator test

sethrj committed:
Add geometry 'inside' check to status executor

sethrj committed:
Do not gather step info for errored tracks

sethrj committed:
Reverse if/unless

sethrj committed:
Loosen tolerance on two tests

sethrj committed:
Final pair code review

sethrj committed:
Loosen one more test

sethrj committed:
Revert "REVERTME: delete other CI builds"

This reverts commit 7956422d081195a2833f4fbc9a9f76e7dc500504.

sethrj committed:
Reduce safety imager output sizes

mrguilima committed:
Merge branch 'develop' into vg-2.x

Not sure that the merge conflits in
.github/workflows/build-spack.yml have been properly fixed.

mrguilima committed:
Fix Geometry tests on the GPU

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vg-2.x

sethrj committed:
Add wildstyle preset

sethrj committed:
Fix GPU test

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into HEAD

sethrj committed:
Wrap vecgeom cleanup with try/catch

sethrj commented:
@mrguilima I think we can merge this after #1652 and https://github.com/spack/spack/pull/49313 are merged! I'll pull those in and update the CI when those are in.

The one outstanding issue is the sensitive behavior of TestEM3 ... but we can disable that for the moment.

sethrj committed:
Add wildstyle surface build

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vg-2.x

mrguilima committed:
Fix using statement

sethrj committed:
Add spack build

sethrj committed:
Add debugging capabilities to heuristic test

sethrj committed:
Work around heuristic test failures for now

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vg-2.x

sethrj committed:
Reapply "REVERTME: delete all but spack CI"

This reverts commit 986166163295592e85d783f09bdc21d8c3a621e0.

sethrj committed:
Revert "Reapply "REVERTME: delete all but spack CI""

This reverts commit 5479586193b601a446b5b8b21f39dbbb30c1d7bf.

sethrj committed:
Fix rocm build failure

sethrj reviewed (approved):
Hooray! Surface crossing issues (when a direction change happens) to be revisited in the next follow-on.
